jplag.frontend.rust/src/main/java/de/jplag/rust/JplagRustListener.java

```diff
@@ -0,0 +1,898 @@
+import de.jplag.rust.grammar.RustParserBaseListener;
+
+public class JplagRustListener extends RustParserBaseListener implements ParseTreeListener {
+
+    private final RustParserAdapter parserAdapter;
+
+    private final ParserState<RustContext> state = new ParserState<>();
+
+    public JplagRustListener(RustParserAdapter parserAdapter) {
+        this.parserAdapter = parserAdapter;
+        state.enter(RustContext.FILE);
+    }
+
+    private void transformToken(int targetType, Token token) {
+        parserAdapter.addToken(targetType, token.getLine(), token.getCharPositionInLine() + 1, token.getText().length());
+    }
+
+    private void transformToken(int targetType, Token start, Token end) {
+        parserAdapter.addToken(targetType, start.getLine(), start.getCharPositionInLine() + 1, end.getStopIndex() - start.getStartIndex() + 1);
+    }
+
+    @Override
+    public void enterInnerAttribute(RustParser.InnerAttributeContext context) {
+        transformToken(INNER_ATTRIBUTE, context.getStart(), context.getStop());
+        super.enterInnerAttribute(context);
+    }
+
+    @Override
+    public void enterOuterAttribute(RustParser.OuterAttributeContext context) {
+        transformToken(OUTER_ATTRIBUTE, context.getStart(), context.getStop());
+        super.enterOuterAttribute(context);
+    }
+
+    @Override
+    public void enterUseDeclaration(RustParser.UseDeclarationContext context) {
+        transformToken(USE_DECLARATION, context.getStart());
+        super.enterUseDeclaration(context);
+    }
+
+    @Override
+    public void enterUseTree(RustParser.UseTreeContext context) {
+        state.enter(RustContext.USE_TREE);
+        super.enterUseTree(context);
+    }
+
+    @Override
+    public void exitUseTree(RustParser.UseTreeContext context) {
+        state.leave(RustContext.USE_TREE);
+        super.exitUseTree(context);
+    }
+
+    @Override
+    public void enterSimplePath(RustParser.SimplePathContext context) {
+        if (state.getCurrent() == RustContext.USE_TREE) {
+            if (context.parent.getChildCount() > 1 && context.parent.getChild(1).getText().equals("::")) {
+                // Not a leaf
+                return;
+            }
+
+            transformToken(USE_ITEM, context.getStart(), context.getStop());
+        }
+        super.enterSimplePath(context);
+    }
+
+    @Override
+    public void enterModule(RustParser.ModuleContext context) {
+        transformToken(MODULE, context.getStart());
+        state.enter(RustContext.MODULE_BODY);
+        super.enterModule(context);
+    }
+
+    @Override
+    public void enterStruct_(RustParser.Struct_Context context) {
+        transformToken(STRUCT, context.getStart());
+        state.enter(RustContext.STRUCT_BODY);
+        super.enterStruct_(context);
+    }
+
+    @Override
+    public void exitStruct_(RustParser.Struct_Context context) {
+        state.leave(RustContext.STRUCT_BODY);
+        super.exitStruct_(context);
+    }
```

Comment:
It is supposed to make it easier to decide which token type to add when encountering symbols that are used in various different contexts. So, instead of
```java
// handle '{'
if (context.getParent() instanceof IfExpressionContext) transform(IF_BODY_START, ...);
else if (context.getParent() instanceof LoopExpressionContext) transform(LOOP_BODY_START, ...);
// many more cases...
```
the code can be changed to
```java
// handle '{'
int tokenType = state.getCurrent().getStartType();
transform(tokenType, ...);
```
and thus, many different uses of the same symbol can be handled concisely.

This worked very well in other frontends already, in this frontend there are more cases and exceptions to consider, so it took a lot more different states to make it work. I tried to document why each of the states seemed necessary to me.